Sources/WendyAgent/Services/ContainerMonitor.swift

```diff
@@ -0,0 +1,312 @@
+                    if parts.count == 2, let maxRetries = Int(parts[1]) {
+                        self = .onFailure(maxRetries: maxRetries)
+                    } else {
+                        self = .onFailure(maxRetries: 5)  // Default to 5 retries
+                    }
+                } else {
+                    self = .no  // Default policy
+                }
+            }
+        }
+    }
+
+    private var containerStates: [String: ContainerState] = [:]
+
+    /// Main monitoring loop
+    func run() async throws {
+        logger.info("Starting container monitor service")
+
+        // Wait for containerd to be ready
+        await waitForContainerd()
+
+        try await withGracefulShutdownHandler {
+            while !Task.isCancelled {
+                do {
+                    try await checkContainers()
+                    try await Task.sleep(nanoseconds: 15_000_000_000)  // Check every 15 seconds
+                } catch {
+                    logger.error("Error in container monitor", metadata: ["error": "\(error)"])
+                    try await Task.sleep(nanoseconds: 30_000_000_000)  // Wait 30 seconds on error
+                }
+            }
+        } onGracefulShutdown: {
+            self.logger.info("Container monitor service shutting down gracefully")
+        }
+
+        logger.info("Container monitor service stopped")
+    }
+
+    /// Wait for containerd to become available
+    private func waitForContainerd() async {
+        let maxAttempts = 30
+        let delayBetweenAttempts: UInt64 = 1_000_000_000
+
+        logger.info("Waiting for containerd to become available...")
+
+        for attempt in 1...maxAttempts {
+            // Check if task is cancelled
+            if Task.isCancelled {
+                logger.info("Container monitor cancelled while waiting for containerd")
+                return
+            }
+
+            do {
+                // Try to connect to containerd
+                _ = try await Containerd.withClient { client in
+                    // List containers to ensure the client is active
+                    _ = try await client.listContainers()
+                    return true
+                }
+
+                logger.info("Successfully connected to containerd after \(attempt) attempt(s)")
+                return
+            } catch {
+                if attempt == 1 {
+                    logger.debug(
+                        "Containerd not ready yet, will retry...",
+                        metadata: ["error": "\(error)"]
+                    )
+                } else if attempt % 5 == 0 {
+                    logger.info(
+                        "Still waiting for containerd (attempt \(attempt)/\(maxAttempts))..."
+                    )
+                }
+
+                // Don't sleep on the last attempt
+                if attempt < maxAttempts {
+                    try? await Task.sleep(nanoseconds: delayBetweenAttempts)
```

Comment:
```suggestion
                    try await Task.sleep(nanoseconds: delayBetweenAttempts)
```

Throwing CancellationError is fine